SR-7732: Dynamic casting CFError to Error results in a memory leak

tbkka opened the pull request:
The special handling for casting CFError/NSError to Swift Error
type was using cleanup code that didn't correctly handle this case.
This replaces the cleanup code with a more targeted version.

Resolves SR-7732
Resolves rdar://problem/40423061

tbkka committed:
SR-7732: Dynamic casting CFError to Error results in a memory leak

The special handling for casting CFError/NSError to Swift Error
type was using cleanup code that didn't correctly handle this case.
This replaces the cleanup code with a more targeted version.

Fixes: SR-7732
Fixes: rdar://problem/40423061

tbkka committed:
Whitespace fixes

DougGregor reviewed (approved):
Thank you! Please cherry-pick to the swift-5.2-branch

benrimmington commented on test/stdlib/ErrorBridged.swift:
```diff
@@ -788,4 +788,23 @@ ErrorBridgingTests.test("error-to-NSObject casts") {
+
+    if let error = something as? Error {
+      expectEqual(
+        "The operation couldn’t be completed. (Foo error 1.)",
```
`"The operation couldn\\U2019t be completed. (%@ error %ld.)"` is [localized][] in CoreFoundation. So will this test fail when the user's preferred language isn't English?

[localized]: <https://github.com/apple/swift-corelibs-foundation/blob/653c070d8ee571611ab4cc3e63d640ef4b8d151c/CoreFoundation/Error.subproj/CFError.c#L307>

tbkka replied:
Good catch!  I did actually realize that would be a problem when I wrote it, but forgot to go back and find a better approach.  Thanks for pointing it out!

tbkka committed:
Don't rely on localizable strings to verify test behavior.

I've verified this simplified test still leaks with
the original code and does not leak with the fixed code.

tbkka commented:
@swift-ci Please test macOS

tbkka commented:
@swift-ci Please smoke test Linux

tbkka committed:
Don't test against old runtimes that predate this fix

tbkka commented:
@swift-ci Please smoke test

tbkka commented:
@swift-ci Please smoke test macOS

tbkka committed:
Explicitly test both NSError and CFError

tbkka commented on test/stdlib/ErrorBridged.swift:
```diff
@@ -788,4 +788,23 @@ ErrorBridgingTests.test("error-to-NSObject casts") {
+// SR-7732: Casting CFError to Error results in a memory leak
+ErrorBridgingTests.test("CFError-to-Error casts") {
+  func should_not_leak() {
+    let something: Any? = NSError(domain: "Foo", code: 1)
```
I added a separate test to explicitly verify CFError.

mikeash replied:
Yay!

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please smoke test macOS
